crates/turborepo-lib/src/package_changes_watcher.rs

```diff
@@ -0,0 +1,279 @@
+use std::collections::HashSet;
+
+use ignore::gitignore::Gitignore;
+use notify::Event;
+use tokio::sync::{broadcast, oneshot, watch};
+use turbopath::{AbsoluteSystemPathBuf, AnchoredSystemPath, AnchoredSystemPathBuf};
+use turborepo_filewatch::{NotifyError, OptionalWatch};
+use turborepo_repository::{
+    change_mapper::{ChangeMapper, GlobalDepsPackageChangeMapper, PackageChanges},
+    package_graph::{PackageGraph, PackageGraphBuilder, PackageName},
+    package_json::PackageJson,
+};
+
+use crate::turbo_json::TurboJson;
+
+pub enum PackageChangeEvent {
+    // We might want to make this just String
+    Package { name: PackageName },
+    Rediscover,
+}
+
+/// Watches for changes to a package's files and directories.
+pub struct PackageChangesWatcher {
+    _exit_tx: oneshot::Sender<()>,
+    _handle: tokio::task::JoinHandle<()>,
+    package_change_events_rx: watch::Receiver<PackageChangeEvent>,
+}
+
+impl PackageChangesWatcher {
+    pub fn new(
+        repo_root: AbsoluteSystemPathBuf,
+        file_events_lazy: OptionalWatch<broadcast::Receiver<Result<Event, NotifyError>>>,
+    ) -> Self {
+        let (exit_tx, exit_rx) = oneshot::channel();
+        let (package_change_events_tx, package_change_events_rx) =
+            watch::channel(PackageChangeEvent::Rediscover);
+        let subscriber = Subscriber::new(repo_root, file_events_lazy, package_change_events_tx);
+
+        let _handle = tokio::spawn(subscriber.watch(exit_rx));
+        Self {
+            _exit_tx: exit_tx,
+            _handle,
+            package_change_events_rx,
+        }
+    }
+
+    pub fn package_changes(&self) -> watch::Receiver<PackageChangeEvent> {
+        self.package_change_events_rx.clone()
+    }
+}
+
+struct Subscriber {
+    file_events_lazy: OptionalWatch<broadcast::Receiver<Result<Event, NotifyError>>>,
+    repo_root: AbsoluteSystemPathBuf,
+    package_change_events_tx: watch::Sender<PackageChangeEvent>,
```

Comment:
Is `watch` the right kind of channel here? IIUC, that doesn't guarantee you'll see every event, just the latest one. I think we probably want a `broadcast` channel.